src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,250 @@ where
+
+        for level in 0..auth_path_rev.len() {
+            // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+            // and the bit being 1 indicates our currently hashed value is on the right.
+            // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+            let bit = path_rev[level];
+            let sibling = auth_path_rev[level];
+
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate_both(
+                two_to_one_hash_parameter,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
+    ///
+    /// Constraints will not be satisfied if root is incorrect.
+    pub fn check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        self.conditionally_check_membership(
+            leaf_hash_parameter,
+            two_to_one_hash_parameter,
+            root,
+            leaf,
+            &Boolean::constant(true),
+        )
+    }
+
+    /// Check membership if `should_enforce` is True.
+    pub fn conditionally_check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+        should_enforce: &Boolean<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        let expected_root =
+            self.calculate_root(leaf_hash_parameter, two_to_one_hash_parameter, leaf)?;
+        // enforce `expected_root` is equal to the given root
+        expected_root.conditional_enforce_equal(root, should_enforce);
+        Ok(())
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, constraints will not be satisfied.
+    pub fn update_leaf(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        old_root: &TwoToOneH::OutputVar,
+        old_leaf: &impl ToBytesGadget<ConstraintF>,
+        new_leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        self.check_membership(
+            leaf_hash_parameter,
+            two_to_one_hash_parameter,
+            old_root,
+            old_leaf,
+        )?;
+        Ok(self.calculate_root(leaf_hash_parameter, two_to_one_hash_parameter, new_leaf)?)
```

Comment:
Doesn't `check_membership` already call `self.calculate_root`? We can reuse the result from that.

Reply:
hmm the parameter passed to `calculate_root `is actually different. In `check_membership`, we passed `old_leaf` to `calculate_root` and in later call we passed `new_leaf` to `calculate_root`

Reply:
btw, when updating the path using `new_leaf`, do we want to check if `old_root` is correct? If not, we can remove `verify_membership` in `update_leaf` so avoid unnecessary constraints. @Pratyush